sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala

```diff
@@ -277,6 +286,34 @@ trait HashJoin extends BaseJoinExec with CodegenSupport {
     }
   }
 
+  private def nullAwareAntiJoin(
+      streamIter: Iterator[InternalRow],
+      hashedRelation: HashedRelation): Iterator[InternalRow] = {
+    if (hashedRelation == EmptyHashedRelation) {
+      streamIter
+    } else if (hashedRelation == EmptyHashedRelationWithAllNullKeys) {
+      // This code branch should only be used in BroadcastHashJoinExec(Non-AQE).
```

Comment:
can we enforce this assumption better than comment? e.g. we can override `nullAwareAntiJoin` in `BroadcastHashJoinExec` with:

```
  protected override def nullAwareAntiJoin(
      streamIter: Iterator[InternalRow],
      hashedRelation: HashedRelation): Iterator[InternalRow] = {
    if (hashedRelation == EmptyHashedRelationWithAllNullKeys &&
        sqlContext.conf.adaptiveExecutionEnabled) {
      Iterator.empty
    } else {
      super.nullAwareAntiJoin(streamIter, hashedRelation)
    }
  }
```

and we can make better assertion at `nullAwareAntiJoin ` in `HashJoin` here:

```
  protected def nullAwareAntiJoin(
      streamIter: Iterator[InternalRow],
      hashedRelation: HashedRelation): Iterator[InternalRow] = {
    if (hashedRelation == EmptyHashedRelation) {
      streamIter
    } else {
      require(hashedRelation != EmptyHashedRelationWithAllNullKeys)
      ...
    }
  }
```

Reply:
Good advice.